lib/semantics/resolve-names.cc

```diff
@@ -3410,25 +3418,45 @@ void DeclarationVisitor::Post(
   for (auto &declaration : x.declarations) {
     auto &bindingName{std::get<parser::Name>(declaration.t)};
     auto &optName{std::get<std::optional<parser::Name>>(declaration.t)};
-    auto &procedureName{optName ? *optName : bindingName};
-    auto *procedure{FindSymbol(procedureName)};
+    const parser::Name &procedureName{optName ? *optName : bindingName};
+    Symbol *procedure{FindSymbol(procedureName)};
     if (!procedure) {
-      Say(procedureName, "Procedure '%s' not found"_err_en_US);
-      continue;
-    }
-    procedure = &procedure->GetUltimate();  // may come from USE
-    if (!CanBeTypeBoundProc(*procedure)) {
-      SayWithDecl(procedureName, *procedure,
-          "'%s' is not a module procedure or external procedure"
-          " with explicit interface"_err_en_US);
-      continue;
+      procedure = NoteInterfaceName(procedureName);
     }
     if (auto *s{MakeTypeSymbol(bindingName, ProcBindingDetails{*procedure})}) {
       SetPassNameOn(*s);
     }
   }
 }
 
+void DeclarationVisitor::CheckBindings(
+    const parser::TypeBoundProcedureStmt::WithoutInterface &tbps) {
+  CHECK(currScope().kind() == Scope::Kind::DerivedType);
```

Comment:
This condition could be written `currScope().IsDerivedType()`.

Reply:
Sure.  And it's probably preferable, in case some future change decides to discriminate between derived types, PDT instantiations, and some future standard feature for generics.  Should `kind()` become private?